dc/s2n-quic-dc/events/acceptor.rs

```diff
@@ -0,0 +1,343 @@
+    local_address: SocketAddress<'a>,
+
+    /// The backlog size
+    backlog: usize,
+}
+
+/// Emitted when a TCP acceptor completes a single iteration of the event loop
+#[event("acceptor:tcp:started")]
+#[subject(endpoint)]
+struct AcceptorTcpLoopIterationCompleted {
+    /// The number of streams that are waiting on initial packets
+    #[measure("pending_streams")]
+    pending_streams: usize,
+
+    /// The amount of time it took to complete the iteration
+    #[timer("processing_duration")]
+    processing_duration: core::time::Duration,
+
+    /// The computed max sojourn time that is allowed for streams
+    ///
+    /// If streams consume more time than this value to initialize, they
+    /// may potentially be replaced by more recent streams.
+    #[measure("max_sojourn_time", Duration)]
+    max_sojourn_time: core::time::Duration,
+}
+
+/// Emitted when a fresh TCP stream is enqueued for processing
+#[event("acceptor:tcp:fresh:enqueued")]
+#[subject(endpoint)]
+struct AcceptorTcpFreshEnqueued<'a> {
+    /// The remote address of the TCP stream
+    #[builder(&'a s2n_quic_core::inet::SocketAddress)]
+    remote_address: SocketAddress<'a>,
+}
+
+/// Emitted when a the TCP acceptor has completed a batch of stream enqueues
+#[event("acceptor:tcp:fresh:batch_completed")]
+#[subject(endpoint)]
+struct AcceptorTcpFreshBatchCompleted {
+    /// The number of fresh TCP streams enqueued in this batch
+    #[measure("enqueued")]
+    enqueued: usize,
+
+    /// The number of fresh TCP streams dropped in this batch due to capacity limits
+    #[measure("dropped")]
+    dropped: usize,
+
+    /// The number of TCP streams that errored in this batch
+    #[measure("errored")]
+    errored: usize,
+}
+
+/// Emitted when a TCP stream has been dropped
+#[event("acceptor:tcp:stream_dropped")]
+#[subject(endpoint)]
+struct AcceptorTcpStreamDropped<'a> {
+    /// The remote address of the TCP stream
+    #[builder(&'a s2n_quic_core::inet::SocketAddress)]
+    remote_address: SocketAddress<'a>,
+
+    #[nominal_counter("reason")]
+    reason: AcceptorTcpStreamDropReason,
+}
+
+enum AcceptorTcpStreamDropReason {
+    /// There were more streams in the TCP backlog than the userspace queue can store
+    FreshQueueAtCapacity,
+
+    /// There are no available slots for processing
+    SlotsAtCapacity,
+}
+
+/// Emitted when a TCP stream has been replaced by another stream
+#[event("acceptor:tcp:stream_replaced")]
+#[subject(endpoint)]
+struct AcceptorTcpStreamReplaced<'a> {
+    /// The remote address of the stream being replaced
+    #[builder(&'a s2n_quic_core::inet::SocketAddress)]
+    remote_address: SocketAddress<'a>,
+
+    /// The amount of time that the stream spent in the accept queue before
+    /// being replaced with another
+    #[timer("sojourn_time")]
+    sojourn_time: core::time::Duration,
+
+    /// The amount of bytes buffered on the stream
+    #[measure("buffer_len", Bytes)]
+    buffer_len: usize,
+}
```

Comment:
It would be good to figure out what questions we want to answer... I think one might be "how close are we to evicting things?"

I think the current metrics will make that hard, maybe we want something like slots_left or some similar measure? Or perhaps a % metric?

Reply:
Yeah I've got `pending_streams` on line 25. I can add another field relating to remaining capacity and percentage.

Reply:
Added in https://github.com/aws/s2n-quic/pull/2379/commits/43e8cb7bf1fb353ee40a7b113e6f38296bcf8f30#diff-9d58bbb85d84818da3d635f23f8c8b06cff3ac114f8a89f51d130250226a7c32R27-R34